### ISSUE TITLE
Add semantics for "expected failures"

### ISSUE THREAD

**Issue by sschuberth** [kotest/kotest#187]
Sometimes, esp. in TDD, it's convenient to add a test that is expected to fail at first, but will be fixed to succeed (e.g. by adding an implementation in the application) later on. In contrast to `config(enabled = false)` these tests *should* be executed but be clearly marked (e.g. in yellow) in the output to document them as "TODOs". The commit adding the implementation then typically also re-configures the test from an expected failure to a regular test that is expected to succeed.

**Comment by sksamuel** [comment]
So if we had an option, todo = true, then the test name could be prefixed with TODO: which would help with TDD style of development?

**Comment by sschuberth** [comment]
Actually, when I was saying "TODO" I was not thinking of TODO-style code comments like `// TODO: Implement this`. I just wanted to give that special "section" in the test report a name where tests that are expected to fail would be listed.

How I believe it should work is that we have `.config(expectedFailure = true)`, and tests marked as such should be run but neither count towards succeeded nor failed tests. Instead, we should amend the summary saying e.g. `1 test completed, 1 failed` with `, 1 expected to fail`.

**Comment by sksamuel** [comment]
The difficulty there is that kotlin test is built on junit 4 and I don't think it has support for "expected to fail" tests. Just failed and completed. So we'd have to write our own test runner framework - and then it wouldn't work with all the tools that already know how to use junit. Hence my idea of having tests that don't run - just automatically pass and have todo prepended to their name automatically.

**Comment by sschuberth** [comment]
I see. Any idea whether first moving to JUnit 5 would make a difference here? It seems that JUnit 5 has been designed to better support TDD, but I yet was unable to find any information about the support of expected failures in JUnit 5.

**Comment by sksamuel** [comment]
In theory we could move to JU5 tomorrow and in fact I think I'll upgrade immediately. Not sure if there's any kind of support for what you want though.

**Comment by helmbold** [comment]
As far as I know JUnit 5 yet, there is no such thing.

**Comment by sksamuel** [comment]
jUnit 5 allows us to specify a reason why a test was skipped. So we could add a message for each of the tests that are expected to fail. In intellij the icon would be the same as it is for any other ignored test, but the custom message might be useful.

**Comment by sschuberth** [comment]
Hmm. The whole point of "expected to fail" tests is that they *do* get run, but their failure does not count towards the fail count. We do not want to skip these tests, thus I'm not sure how a custom reason why a test was skipped helps here.

**Comment by sksamuel** [comment]
You can run them, and then pretend they were skipped if they fail.

**Comment by sschuberth** [comment]
Ah, so that's indeed an option then!

**Comment by sksamuel** [comment]
I'd like to add this to 3.2 in the form of pending tests. KotlinTest can track the number of pending tests then output this at the end of the build. Pending tests will execute, but will not fail the test - unless they pass. A successful pending test should be considered an error .

**Comment by LeoColman** [comment]
Bumping this.

Do we still want to implement this?
If so, what do we want to do?

**Comment by BorzdeG** [comment]
Why it is impossible to use inside the "shouldThrow" test?

**Comment by LeoColman** [comment]
We also have `shouldFail`, which is specific to AssertionErrors...
I think this should be fine. 
I'll close this guy. Should this need rise again, we can reopen it
@sksamuel

**Comment by sksamuel** [comment]
I think the idea is that you mark tests as pending because you haven't
implemented the feature yet.

Then once you inplement the feature the test starts to pass and so the test
framework fails it.

I'm not sure of the benefit to this approach myself but other test
frameworks have it.

**Comment by LeoColman** [comment]
I think that shouldFail can accomplish the same thing, but we can implement a dedicated feature for this id you think it's worth it

**Comment by sksamuel** [comment]
Yes I guess it can although it doesn't get included in a separate column in
the output. But not a huge deal is it.

**Comment by sksamuel** [comment]
https://etorreborre.github.io/specs2/guide/SPECS2-4.3.4/org.specs2.guide.PendingUntilFixedExamples.html

**Comment by dajulia3** [comment]
Just throwing my two cents in, my team and I would really like to see "pending" with separate output, especially in CI.

**Comment by brucek** [comment]
For what it's worth, my team is hitting the need for this capability as well.

The scenario:
- we have a known code defect
- we don't have time to actually fix it now
- we can create a test case indicating the failure

...if we had the equivalent of `pendingUntilFixed` ([rspec](https://relishapp.com/rspec/rspec-core/v/2-0/docs/pending/pending-examples) also does a good job of this) we can note the item to fix in the code, and be reminded of all such pending tech debt items every time we run the test suite.

It is also helpful in that if some other code causes the test to unexpectedly pass, I am notified (by a test suite failure), so I can go investigate and see if something unexpected happened, if it fixed the issue, etc.

Otherwise if I am not notified of the test passing immediately as a test failure, someone later might run the test, see it pass and assume it was properly fixed (which it may not have been).

**Comment by sschuberth** [comment]
@brucek, I've reopened this and copied your comment from https://github.com/kotest/kotest/issues/3061 here:

> I couldn't re-open the issue in #187 but this does have value, esp coming to Kotlin from other more developed ecosystems, so I'd like to reopen and reconsider this - key information copied here:
> 
> [sschuberth](https://github.com/sschuberth) Sometimes, esp. in TDD, it's convenient to add a test that is expected to fail at first, but will be fixed to succeed (e.g. by adding an implementation in the application) later on. In contrast to config(enabled = false) these tests should be executed but be clearly marked (e.g. in yellow) in the output to document them as "TODOs". The commit adding the implementation then typically also re-configures the test from an expected failure to a regular test that is expected to succeed.
> 
> [brucek](https://github.com/brucek) For what it's worth, my team is hitting the need for this capability as well.
> 
> The scenario:
> 
> we have a known code defect we don't have time to actually fix it now we can create a test case indicating the failure ...if we had the equivalent of pendingUntilFixed ([rspec](https://relishapp.com/rspec/rspec-core/v/2-0/docs/pending/pending-examples) also does a good job of this) we can note the item to fix in the code, and be reminded of all such pending tech debt items every time we run the test suite.
> 
> It is also helpful in that if some other code causes the test to unexpectedly pass, I am notified (by a test suite failure), so I can go investigate and see if something unexpected happened, if it fixed the issue, etc.
> 
> Otherwise if I am not notified of the test passing immediately as a test failure, someone later might run the test, see it pass and assume it was properly fixed (which it may not have been).

**Comment by sksamuel** [comment]
So what I don't understand is what differs from a regular test. So your three options are - failed test, successful test, ignored test. Which of those is the desired output? Are you wanting to make a test that's empty but fails?

**Comment by sschuberth** [comment]
> So your three options are - failed test, successful test, ignored test.

I was hoping there was the possibility to add a fourth option, like "pending test" or "expected failure test".

If that's not possible I'd say it should either:
- be a failed test that does not could against the Gradle task result to be a failure
- be a successful test that uses some mechanism (like console output or special test naming) to indicate that it actually documents the presence of a bug
- be an ignored test (i.e. its icon in IntelliJ is that of an ignored test) that actually *does* get executed (and thus is not really ignored)

**Comment by sksamuel** [comment]
As part of Kotest 6.0 I hope to ditch JUnit but realistically the JVM world revolves around that. That's why this feature has always been a bit "impossible".

The first can't happen, the 2nd is possible - add some kind of **** PENDING ***** to the name or something. I think the third could happen, but it would be a lot of work - execute a test first and then decide if it was 'ignored' later. Due to the way intellij and gradle work, the third is problematic.

**Comment by sschuberth** [comment]
> So what I don't understand is what differs from a regular test.

Coming back to that, the difference is basically just semantics. Of course you could write an "expected to fail"-test in such a way that it actually succeeds by expecting the wrong result. The problem with this is that you can't distinguish such a test from a "real" test anymore; but in fact you'd like the "expected to fail"-test to remind you of what still needs to be fixed. Think of it as a programmatic replacement for writing human-readable issue reports.

So ideally, you'd even be able to run all "expected to fail"-tests in your code base at once. While writing this, it just came to my mind that this could maybe be addressed via a special test tag. A built-in "expectedFailure"-tag would allow to run all such tests at once, and Kotest could contain some logic to invert the assertions in such tests.

Then, once the fix for the failure is committed, all you'd need to do in the test would be to remove the "expectedFailure"-tag, and everything should pass.

**Comment by sksamuel** [comment]
That makes sense:

```kotlin
test("foo").config(pending = true) { 
 //
}
```

And this is inverted - if it passes it fails and vice versa.

**Comment by sschuberth** [comment]
> And this is inverted - if it passes it fails and vice versa.

Right, that's similar to what I wrote [above](https://github.com/kotest/kotest/issues/187#issuecomment-335379031). It just would be nice if such pending tests could also be marked in a special way in console output and test reports. And, like said, if you could run all of them in a single blow. Maybe `config(pending = true)` could imply a tag.

**Comment by sksamuel** [comment]
`pending = true` could easily:
* imply a tag (if we want to do that?)
* add ***PENDING*** to output (and in Kotest 6 be highlighted properly)
* invert the result

**Comment by mikezx6r** [comment]
Another feature/scenario could be 'wip' tests. Check out this library for writing Domain Driven Testing. https://github.com/uberto/pesticide

It's a custom framework, written on top of JUnit's test generation features.

Anyhow, to my point. It is built on the premise that you'll write your Acceptance test FIRST, and mark it as 'wip', where 'wip' takes a LocalDate, and a String. The LocalDate is the 'expected to fail until date', and the String is basically a 'comment'.

If the test runs, and it's before the date, then it inverts the result. Otherwise, it reports the result as-is.

It uses 'Ignored' as the marker.

So a minor enhancement to your existing proposal, but reduces/negates the need for the team to review the test output, looking for PENDING items.

Complete side note. I found the approach presented in Uberto's book very interesting. Worth checking out the book, or at least the sample code IMO. Not affiliated with him, or the book.

**Comment by sschuberth** [comment]
Let's keep this open for something to be considered for Kotest 6.

**Comment by zakhenry** [comment]
Partially bumping the thread, but also offering an alternative use case for this desire.

Recently I encountered a bug in a github repo, and I reported the issue in a way that allowed one of their devs to immediately write a unit test, and confirm that my reproduction was good with the latest version of the code. They flagged that test as expecting failure, and merged it into the codebase.

As a user of the framework it felt like a really good user experience to see that my bug was validated, and knew that in future it would not be forgotten about because there was a unit test in there telling them that something was broken and it would never get stale.

edit: here's the issue in question, see how there was two separate PRs, the first one validated my repro and the second actually fixed it a few weeks later https://github.com/microsoft/playwright/issues/17824

**Comment by sschuberth** [comment]
> but also offering an alternative use case for this desire.

That's actually exactly the use-case I had in mind initially.

**Comment by sksamuel** [comment]
Closing as this not possible (as I'm cleaning up tickets), as until JUnit platform adds support, there's is nothing we can do really.
And JUnit platform 6 just came out so it's not likely to be added any time soon (ever?)
